dislib/classification/rf/decision_tree.py

```diff
@@ -215,213 +176,245 @@ def compute_split(tree_path, sample, n_features, features_mmap, y_s, n_classes,
-                                                       b_value)
+        groups = get_groups(sample, y_s, features_mmap, b_index, b_value)
+        left_group, y_l, right_group, y_r = groups
         if left_group.size and right_group.size:
             split_ended = True
-            node = InternalNode(tree_path, b_index, b_value)
+            node_info = InnerNodeInfo(b_index, b_value)
         else:
             tried_indices.extend(list(index_selection))
             if len(tried_indices) == n_features:
                 split_ended = True
-                node = build_leaf(y_s, tree_path)
+                node_info = compute_leaf_info(y_s, n_classes)
 
-    return node, left_group, y_l, right_group, y_r
+    return node_info, left_group, y_l, right_group, y_r
 
 
-def flush_remove_nodes(file_out, nodes_to_persist):
-    flush_nodes(file_out, *nodes_to_persist)
-    for obj in nodes_to_persist:
-        compss_delete_object(obj)
-    del nodes_to_persist[:]
-
-
-@task(file_out=FILE_INOUT)
-def flush_nodes(file_out, *nodes_to_persist):
-    with open(file_out, "a") as tree_file:
-        for item in nodes_to_persist:
-            if isinstance(item, (Leaf, InternalNode)):
-                tree_file.write(item.to_json())
-            else:
-                for nested_item in item:
-                    if isinstance(nested_item, TreeWrapper):
-                        nested_item.write_to(tree_file)
-                    else:
-                        tree_file.write(nested_item.to_json())
-
-
-def build_subtree(sample, y_s, n_features, tree_path, max_depth, n_classes,
-                  m_try, samples_file, features_file):
+def build_subtree(sample, y_s, n_features, max_depth, n_classes, m_try,
+                  samples_file, features_file):
     if features_file is not None:
-        return build_subtree_a(sample, y_s, n_features, tree_path, max_depth,
-                               n_classes, m_try, samples_file,
-                               features_file)
+        return build_subtree_a(sample, y_s, n_features, max_depth, n_classes,
+                               m_try, samples_file, features_file)
     else:
-        return build_subtree_b(sample, y_s, n_features, tree_path, max_depth,
-                               n_classes, m_try, samples_file)
+        return build_subtree_b(sample, y_s, n_features, max_depth, n_classes,
+                               m_try, samples_file)
 
 
-@task(samples_file=FILE_IN, features_file=FILE_IN, returns=list)
-def build_subtree_a(sample, y_s, n_features, tree_path, max_depth, n_classes,
-                    m_try, samples_file, features_file):
-    return build_subtree_in(sample, y_s, n_features, tree_path, max_depth,
-                            n_classes, m_try, samples_file,
-                            features_file=features_file)
+@task(samples_file=FILE_IN, features_file=FILE_IN, returns=Node)
+def build_subtree_a(sample, y_s, n_features, max_depth, n_classes, m_try,
```

Comment:
Same as with split_node_a and b

Reply:
Same as before. I've renamed the tasks to _build_subtree, _build_subtree_with_features, the containing function _build_subtree_wrapper, and a contained function _compute_build_subtree.